File: tools/deadprotos.py
#!/usr/bin/env python3
"""Delete prototypes for functions that no longer exist.

Usage:
    python3 tools/deadprotos.py <vim.c>

The dead-code sweep keys on `-Wall`, and gcc says nothing at all about a
declaration of a function that is never defined and never called: it is a
promise about some other translation unit, and in a file that IS the whole
program that promise is simply false.

They matter because of what they keep alive.  `proftime_T` and `getoption_T`
survived a whole type sweep -- correctly, by that sweep's own rule -- because
each was still mentioned by one prototype, and a prototype is a root.  Delete
the prototype and the type becomes unreachable, which is why this runs inside
the same fixpoint loop rather than once.

The test is exact and needs no C parsing: a declaration at file scope whose
name appears NOWHERE else in the file.  A defined function's name appears at
least twice (its prototype and its definition), a called one more, so one
occurrence means nothing refers to it.

**Everywhere except osdef.h and xdiff.h.**  A prototype in this program's own
headers naming a function that does not exist is simply wrong.  Those two
blocks are not: osdef.h declares libc functions the platform may not have
declared, and xdiff.h declares the xdiff library's interface -- both describe
code that lives somewhere else by design, and deleting them because nothing
here calls them would be deleting a true statement.  Measured: the reference
vim.c keeps _Xmblen and all five xdl_* declarations, and keeps mmfile_t and
xpparam_t alive through them, while dropping alloc_id, proftime_time_left and
get_option_value from vim's own headers.
"""

import re
import sys
from collections import Counter
from pathlib import Path

# A prototype at file scope: type, name, parameter list, semicolon, all on one
# line -- which Phase 7 guarantees, having joined every parenthesised group.
PROTO = re.compile(r'^[A-Za-z_][A-Za-z0-9_ \t*]*?\b(\w+)\s*\([^;]*\)\s*;\s*$')
WORD = re.compile(r'[A-Za-z_]\w*')
DEFINE = re.compile(r'^[ \t]*#[ \t]*define[ \t]+(\w+)')


def main():
    if len(sys.argv) != 2:
        sys.exit(__doc__)
    path = Path(sys.argv[1])
    lines = path.read_text(errors='surrogateescape').split('\n')

    # A mention inside a #define body counts only if the MACRO is used.  An
    # unused macro's body is not a reference to anything -- alloc_id's only
    # other mention is inside ALLOC_ONE_ID, which nothing calls, and Phase 9
    # deletes 910 macros for exactly that reason.
    #
    # But ignoring every macro body is wrong in the other direction, and gcc
    # says so immediately: exestack and au_new_curbuf are named only from macro
    # bodies, and those macros are called all over the file.  So: live macros
    # contribute their bodies, dead ones do not, and liveness is itself a
    # fixpoint because a live macro can be what makes another one live.
    outside = Counter()
    defines = []
    for line in lines:
        m = DEFINE.match(line)
        if m:
            defines.append((m.group(1), line))
            continue
        outside.update(WORD.findall(line))

    seen = Counter(outside)
    live = set()
    while True:
        grew = False
        for name, line in defines:
            if name not in live and seen[name] > 0:
                live.add(name)
                seen.update(WORD.findall(line))
                grew = True
        if not grew:
            break

    KEEP = ('osdef.h', 'xdiff.h')
    protected = set()
    inside = None
    for i, line in enumerate(lines):
        if line.startswith('// ') and 'begin ' in line:
            inside = next((k for k in KEEP if 'begin ' + k in line), None)
        elif line.startswith('// ') and 'end ' in line and inside:
            if 'end ' + inside in line:
                inside = None
        if inside:
            protected.add(i)

    out = []
    dropped = []
    for i, line in enumerate(lines):
        m = PROTO.match(line) if i not in protected else None
        if m and seen[m.group(1)] == 1:
            dropped.append(m.group(1))
            continue
        out.append(line)

    if dropped:
        path.write_text('\n'.join(out), errors='surrogateescape')
    print('  deadprotos   %d declared, never defined, never called%s'
          % (len(dropped),
             ': ' + ', '.join(dropped[:4]) + ('...' if len(dropped) > 4 else '')
             if dropped else ''))


if __name__ == '__main__':
    main()

File: tools/go/internal/dead/deadprotos.go
// Package dead holds the six dead-code tools tools/sweep.sh runs before
// canon, in its order: deadsweep, deadprotos, typereach, funcreach,
// deadfields, deadenums.
//
// They feed each other -- deleting a function orphans a type, deleting a type
// orphans a prototype, deleting a field orphans an enumerator -- so none is
// finished until all are, which is why sweep.sh loops them to a fixpoint on
// the file's digest rather than running each once.
//
// Unlike the canonicalisers, these do not all rewrite unconditionally and do
// not all normalise a trailing newline.  Which is which is part of the
// contract: tools/sweep.sh sees only the file's sha256, so a tool that adds a
// newline the Python would not adds a round to every sweep.
package dead

import (
	"bytes"
	"regexp"
)

// proto is a prototype at file scope: type, name, parameter list, semicolon,
// all on one line -- which the canonicalisers guarantee, having joined every
// parenthesised group onto one line.
var proto = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_ \t*]*?\b(\w+)\s*\([^;]*\)\s*;\s*$`)

var word = regexp.MustCompile(`[A-Za-z_]\w*`)

var define = regexp.MustCompile(`^[ \t]*#[ \t]*define[ \t]+(\w+)`)

// keepBanners are the two former headers whose prototypes must never be
// dropped.  A prototype in this program's own headers naming a function that
// does not exist is simply wrong; these two are not.  osdef.h declares libc
// functions the platform may not have declared and xdiff.h declares the xdiff
// library's interface -- both describe code that lives elsewhere by design,
// and deleting them because nothing here calls them would delete a true
// statement.
var keepBanners = []string{"osdef.h", "xdiff.h"}

// DeadProtos deletes prototypes for functions that no longer exist, and
// returns the names it dropped, in file order.
//
// The dead-code sweep keys on -Wall, and gcc says nothing at all about a
// declaration of a function that is never defined and never called: it is a
// promise about another translation unit, and in a file that IS the whole
// program that promise is false.  They matter for what they keep alive -- a
// prototype is a root for the type sweep, so deleting one can make a type
// unreachable, which is why this runs inside the fixpoint and not once.
//
// The test needs no C parsing: a declaration at file scope whose name appears
// NOWHERE else in the file.  A defined function's name appears at least twice,
// its prototype and its definition; a called one more.  One occurrence means
// nothing refers to it.
func DeadProtos(src []byte) (out []byte, dropped [][]byte) {
	lines := bytes.Split(src, []byte{'\n'})

	// A mention inside a #define body counts only if the MACRO is used.  An
	// unused macro's body is not a reference to anything.  But ignoring every
	// macro body is wrong in the other direction -- some names are mentioned
	// only from macro bodies whose macros are called all over the file -- so
	// live macros contribute their bodies and dead ones do not, and liveness
	// is itself a fixpoint, because a live macro can be what makes another
	// one live.
	//
	// Dead for whim and zero, which have no #define at all; 34,330 captures
	// and 24,385 fixpoint steps on slim, measured.
	seen := map[string]int{}
	type macro struct {
		name string
		line []byte
	}
	var defines []macro
	for _, line := range lines {
		if m := define.FindSubmatch(line); m != nil {
			defines = append(defines, macro{string(m[1]), line})
			continue
		}
		for _, w := range word.FindAll(line, -1) {
			seen[string(w)]++
		}
	}
	live := map[string]bool{}
	for {
		grew := false
		for _, d := range defines {
			if !live[d.name] && seen[d.name] > 0 {
				live[d.name] = true
				for _, w := range word.FindAll(d.line, -1) {
					seen[string(w)]++
				}
				grew = true
			}
		}
		if !grew {
			break
		}
	}

	protected := map[int]bool{}
	inside := ""
	for i, line := range lines {
		if bytes.HasPrefix(line, []byte("// ")) && bytes.Contains(line, []byte("begin ")) {
			inside = ""
			for _, k := range keepBanners {
				if bytes.Contains(line, []byte("begin "+k)) {
					inside = k
					break
				}
			}
		} else if bytes.HasPrefix(line, []byte("// ")) && bytes.Contains(line, []byte("end ")) &&
			inside != "" {
			if bytes.Contains(line, []byte("end "+inside)) {
				inside = ""
			}
		}
		if inside != "" {
			protected[i] = true
		}
	}

	kept := make([][]byte, 0, len(lines))
	for i, line := range lines {
		if !protected[i] {
			if m := proto.FindSubmatch(line); m != nil && seen[string(m[1])] == 1 {
				dropped = append(dropped, m[1])
				continue
			}
		}
		kept = append(kept, line)
	}

	// No trailing-newline normalisation, and the caller writes only when
	// something was dropped: both are this tool's contract and not an
	// oversight.  Joining the fields split on '\n' reproduces the input
	// exactly when nothing is removed.
	return bytes.Join(kept, []byte{'\n'}), dropped
}
